### PR TITLE
First step of code  expr

### PR DESCRIPTION
**Pull request by peter1000**
This would be the first step of the :code expre

I still have some dummy in `macros.jl` as I need to check what exactely should be passed - maybe you know?

Anyway this should do it for a start to get things going on Lexicon and see how that could work out.

I will use for a test this function
```
"function `myrehash` with paramter types."
function myrehash{K,V}(h::Dict{K,V}, k::ByteString, newsz=length(h.keys), x::Dict{K,V}=Dict([(:con, 10)]); 
                                            rm::Bool=false, prifix_ext::Dict{K,V}=Dict([(:remote, false)]))
end

myrehash (generic function with 3 methods)
```

### CODE REVIEW

File: src/builddocs.jl [hunk 1]
```diff
@@ -99,13 +99,14 @@ end
...
 
     source = (linenumber(line), file)
     docs   = Docs{meta.data[:format]}(exec(state, mac))
 
     object = Comment()
-    entry  = Entry{:comment}(meta.modname, source, docs)
+    entry  = Entry{:comment}(meta.modname, source, code, docs)
 
     object, entry
 end
```
Comment (MichaelHatherly) [inline src/builddocs.jl]: `@comment` shouldn't have any code associated with it.

Comment (peter1000) [inline src/builddocs.jl]: just a question : why not  - would it not be better to have one unified Entry

**Comment by MichaelHatherly** [comment]
We should only be collecting code from method bodies and macro bodies rather than from every documented object.

So the categories should only be `:method` and `:macro`, but not a `:tuple` or `:vcat` type of docstring since they also produce a `:method` category.

File: src/builddocs.jl (hunk 1 above)
Comment (MichaelHatherly) [inline src/builddocs.jl]: `@comment` takes the form:

```jl
@comment """

"""
```

There is no associated code with it.

**Comment by peter1000** [comment]
what about types ect?

File: src/builddocs.jl (hunk 1 above)
Comment (MichaelHatherly) [inline src/builddocs.jl]: I'm playing around with some experimental things (may have relevance to https://github.com/MichaelHatherly/Lexicon.jl/issues/71, ie document without having to import Docile), so we might be able to unify some internals.

**Comment by MichaelHatherly** [comment]
> what about types ect?

We could do, there's not much to store though since we can get fieldnames and fieldtypes via reflection on the Lexicon side without `:code` storage.

File: src/builddocs.jl (hunk 1 above)
Comment (peter1000) [inline src/builddocs.jl]: That would be really great in many respects: probably one could get rid of the internal @doc and just read the """docstring""" and leave the macro side for specials like @doc+

Comment (peter1000) [inline src/builddocs.jl]: at the moment comments are associated with the next code if one looks through the EXPR. see: #86

Comment (MichaelHatherly) [inline src/builddocs.jl]: Watch out with the unquoted macro names :) I'm sure doc doesn't want to be pinged on this discussion, sorry about that doc.

Comment (MichaelHatherly) [inline src/builddocs.jl]: `@doc+` is covered by:

```jl
"foobar docs"
foobar
```
or
```jl
"foobar docs"
:foobar
```

**Comment by peter1000** [comment]
adds :code to Docile entries

I think like that we can enable disable at one place which kind of code
expr are added.

**Comment by peter1000** [comment]
> We could do, there's not much to store though since we can get fieldnames and fieldtypes via reflection on the Lexicon side without :code storage.

Just to make sure I understand correctly.  You mean in Lexicon do tosomething like

```
parsedobj = parse("$(obj)")
```
or
```
parsedobj = parse( "begin $obj end")
```

taken from Docile how you do: 
```
parsefile(file) = parse("begin $(readall(file)) end")
```

would you general prefer to do something in Lexicon instead of passing the :source - but that would mean douple parsing?

**Comment by MichaelHatherly** [comment]
The only information we'd *possibly* want from a type is fieldnames, fieldtypes and parameters which can all be found using things like:

```
julia> fieldnames(Rational)
2-element Array{Symbol,1}:
 :num
 :den

julia> fieldtype(Rational{Int}, :num)
T<:Integer

julia> Rational.parameters
svec(T<:Integer)

```

by investigating the `obj` that's stored. So no need for parsing anything for this.

Internal constructors can be documented separately, so the `Expr` for a type with internal constructors isn't needed as far as I'm concerned.

File: src/types.jl [hunk 2]
```diff
@@ -149,14 +150,30 @@ end
...
+"Adds and filters which code are added to entry.data"
+function addentry_codedata!(entry::Entry, code)
+    if (ismethod(code)  ||
+        ismacro(code)   ||
+        istype(code)    ||
+        isalias(code)   ||
+        isglobal(code)  ||
+        issymbol(code)  ||
+        isquote(code)
+        )
```
Comment (MichaelHatherly) [inline src/types.jl]: I'd suggest only capturing `method` and `macro` for the moment. `alias`, `global`, `symbol`, and `quote` don't give us anything useful. Reasons for not capturing `type` I've mentioned in the main comments as a reply.

**Comment by MichaelHatherly** [comment]
> BTW: do you use `julia` a lot outside of Docile/Lexicon too - did not see many repos

Nothing public currently. I wouldn't have written it if it wasn't going to benefit me though ;)

**Comment by peter1000** [comment]
ok I change that - if needed we can always easily add it.

Making some procress on Lexicon kwarg ect.

Question: do you thing it would make sense to have a Config option to output signature and such without spaces like this

```
function convert{P<:Ptr,T<:Ptr}(::Union(Type{Ptr{P}},Type{Ref{P}}),a::Array{T})
end
```

**Comment by MichaelHatherly** [comment]
> do you thing it would make sense to have a Config option to output signature and such without spaces

Not for the moment I think. The spaces allow longer signatures to split across lines a bit better I've found. If it's not much of a change then give it a go if you'd like.

**Comment by peter1000** [comment]
I hope the PR is now ok

**Comment by MichaelHatherly** [comment]
Thanks.

**Pull request merged**